mkosi/_version.py

```diff
@@ -0,0 +1,32 @@
+# SPDX-License-Identifier: LGPL-2.1-or-later
+
+import datetime
+import os
+import subprocess
+from importlib.metadata import PackageNotFoundError, version
+from pathlib import Path
+
+try:
+    __version__ = version("package-name")
+except PackageNotFoundError:
+    # This branch means, that mkosi was not installed as a package as Python is aware, this is most likely
+    # the case if mkosi was run via the bin/mkosi shim in mkosi's repository. Though discouraged, we fall
+    # back to importing setuptools_scm directly and having it try figure stuff out.
+    try:
+        p = subprocess.run(
+            ["git", "describe"],
+            cwd=Path(__file__).parent.parent,
+            check=True,
+            text=True,
+            capture_output=True,
+        )
+        version, *rest = p.stdout.strip().split("-")
+        version = version.lstrip("v")
+        if rest:
+            numcommits, commit = rest
+            __version__ = f"{version}.post1.dev{numcommits}+{commit}.d{datetime.datetime.now():%Y%m%d}"
+        else:
+            # we are exactly on a trag
+            __version__ = version
+    except (subprocess.CalledProcessError, NotADirectoryError):
+        __version__ = os.getenv("MKOSI_VERSION", "0")
```

Comment:
Maybe we should log a warning if we fall back to 0 here?

Reply:
I intended the 0 to be a warning of some sort, since it's a mostly a non-sensical value. That there is no warning at the moment is a feature to support the hack when generating a zipapp. In one local iteration I had changed the installation of the zipapp during build to pip, but this doesn't make the version available, so now the version just get's patched in rather ham-fistedly. if there were a warning here, 

I'm not opposed to adding war warning, I just don't know how to best add one (or make the current hack even more unpalatable).

Reply:
Made this slightly less ham-fisted to also support plain tar balls and it now it also warns for the ultimate fallback.